launch configs as preferences

akosyakov opened the pull request:
Includes:
- launch configurations as preferences
- compatibility with VS Code configuration files, fix https://github.com/theia-ide/theia/issues/2440
  - [over 800 compatibility tests](https://github.com/theia-ide/theia/blob/ak/laung_preferences/packages/debug/src/browser/preferences/launch-preferences.spec.ts)
- ~compound launch configurations~ i will completely remove it from this PR and Theia UI, should be implemented separately - https://github.com/theia-ide/theia/issues/5008
- close #4743 (superseded), fix #2290

Running VS Code tests:
- check out https://github.com/akosyakov/vscode-launch under plugins and build `yarn && yarn compile`
- add the following configuration to launch.json:
```json
    {
      "name": "Launch VS Code Tests",
      "type": "node",
      "request": "launch",
      "args": [
        "${workspaceFolder}/examples/browser/src-gen/backend/main.js",
        "${workspaceFolder}/plugins/vscode-launch/tests",
        "--port",
        "3030",
        "--hostname",
        "0.0.0.0",
        "--extensionTestsPath=${workspaceFolder}/plugins/vscode-launch/out/test",
        "--hosted-plugin-inspect=9339"
      ],
      "env": {
        "THEIA_DEFAULT_PLUGINS": "local-dir:${workspaceFolder}/plugins"
      },
      "stopOnEntry": false,
      "sourceMaps": true,
      "outFiles": [
        "${workspaceFolder}/../.js"
      ]
    }
```
- run the configuration in no debug mode and open browser

TODO:
- [x] testing via VS Code extension - https://github.com/akosyakov/vscode-launch
- [x] unit testing of update configurations
- [x] unit testing of delete configurations
- [x] improve unit tests robustness, e.g. don't rely on file watching
- [x] testing in multi-root workspace context
  - [x] testing preference widget
  - [x] testing preference content assist
- [x] verify https://github.com/theia-ide/theia/issues/4188
- [x] verify https://github.com/theia-ide/theia/pull/4947#issuecomment-484471437
- [x] verify https://github.com/theia-ide/theia/issues/4912#issue-433159593
- [x] verify https://github.com/theia-ide/theia/issues/4773

benoitf commented on packages/core/src/common/preferences/preference-scope.ts:
```diff
@@ -57,7 +57,7 @@ export namespace PreferenceScope {
                 return PreferenceScope.User;
             case 'window':
-                return PreferenceScope.Workspace;
+                return PreferenceScope.Folder;
```
hello, in which case do we have preferencescope.workpace ?

akosyakov replied:
it is inclusive, i.e folder allows workspace and global

elaihau replied:
does it mean, users wouldn't be able to define prefs only available at the "workspace" & "user" level?

akosyakov replied:
actually the whole scope is valid check is bogus right now, it should be aligned with: 
https://github.com/Microsoft/vscode/blob/750680d2f7ae563529501f8ec419d83ccf8ec325/src/vs/workbench/services/configuration/common/configuration.ts#L23-L26

It is not inclusive,  folder preferences allow ONLY resource, workspace preference allow ONLY  window and resource level. I will rework it as well.

akosyakov replied:
@elaihau @benoitf Would be fine to fix it with a follow-up PR? otherwise it would be postponed more and it does not seem to relate to referenced issues.

benoitf replied:
yes

elaihau replied:
yes it could be resolved in a separate PR. since the next release would be either on Apr 30 or May 1, do we want to hold this one until the release is made?

akosyakov replied:
please ask on Spectrum, if someone has concrete objections, yes

from my side, most important that it should be tested not only by me, but double checked by someone else

benoitf commented:
I tried this PR + backporting on this branch https://github.com/theia-ide/theia/commit/ed40d907c4c563573ee475c8f720c2e7df9d6e4e

But I got
```
peError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type undefined
    at assertPath (path.js:39:11)
    at Object.join (path.js:1152:7)
    at Config.diagramsRoot (/private/var/folders/tg/_5rxbhmj4xncz4szvpgswrmc0000gn/T/vscode-unpacked/jebbs.plantuml-2.10.9.vsix/extension/out/src/plantuml/config.js:67:27)
```

FYI source code is
```typescript
diagramsRoot(uri) {
        let folder = uri ? vscode.workspace.getWorkspaceFolder(uri) : undefined;
        if (!folder)
                 return undefined;
LINE67       --> let fsPath = path.join(folder.uri.fsPath, this.read("diagramsRoot", uri));
                 return vscode.Uri.file(fsPath);
    }

```

and read method is there
https://github.com/qjebbs/vscode-plantuml/blob/master/src/plantuml/configReader.ts#L48-L73

akosyakov commented:
@benoitf it is WIP

akosyakov commented:
@elaihau It is not finished, but could you test support of preferences in multi-root workspace context with this PR?

tsmaeder commented:
@akosyakov would https://github.com/theia-ide/theia/issues/4773 be fixed by this PR, as well?

akosyakov commented:
@tsmaeder I hope this PR brings preference impl as close as possible to VS Code expectations, so it should be covered.

akosyakov commented:
oh, finally all compatibility tests from from VS Code extension are running:
<img width="2032" alt="Screen Shot 2019-04-25 at 16 15 28" src="https://user-images.githubusercontent.com/3082655/56742657-726b9b00-6775-11e9-9b35-7e6ae460adc2.png">

I will ditch support of compound debug configuration for this PR next, [fix scope validation](https://github.com/theia-ide/theia/pull/4947#discussion_r278554345) and start verifying whether issues are resolved with this.

@akurinnoy I've changed a bit of code for the plugin system integration. There were 2 issues:
-  a race condition between update and events, sometimes after resolving update, fresh preferences were not available yet, now update waits till changes are propagated
- handling of undefined, there were checks using `!` so some values (0, null, '') were swallowed, i've replaced it with explicit `undefined` check

Other changes are only to core, preferences and debug extensions:
- There is a new concept as a preference configuration. The main config points to settings file, other points to extra settings files which can be embedded in the main, right now it is `launch.json`, next we should rework the task extension cc @elaihau 
- Debug extension does not parse anymore `launch.json` itself but relies on preferences for it. Again the same should be done for the task extension. I will open a follow-up issue for it.
- Special handling for the launch section is encapsulated in the debug extension, not in the preference extension like before.

akosyakov commented:
When i test #4188 i see that launch configuration are created properly and without errors now, but i cannot run them. Should i investigate further as part of this PR or we do it in next step. @akurinnoy @svor ? I'm using the latest java and java-debug vscode extensions with app from https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html

akosyakov commented:
@ariel-bentu please check whether your launch configuration tests are passing with this PR, they do for me

akosyakov commented:
@benoitf what should I do to reproduce the issue with plantuml extension? I've tried to export to svg and it worked out without any issues

akosyakov commented:
@tsmaeder Could you verify please whether #4773 is resolved with this?

akosyakov commented:
@elaihau a follow-up issue for tasks: https://github.com/theia-ide/theia/issues/5013 please test preference support in multi-root workspace context when you have time

benoitf commented:
@akosyakov well if it works then I would say it's fine 
you may check that at these lines
https://github.com/qjebbs/vscode-plantuml/blob/master/src/plantuml/configReader.ts#L48-L73

it returns a value but I would say it is

akosyakov commented:
@benoitf I see, i think it was caused by converting empty string to undefined, now it should remain empty string

svor commented:
> When i test #4188 i see that launch configuration are created properly and without errors now, but i cannot run them. I'm not sure whether it is related to preferences. Should i investigate further as part of this PR or we do it in next step @akurinnoy @svor ? I'm using the latest java and java-debug vscode extensions with app from https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html

@akosyakov Could you try it with java-debug-0.16.0. I think it should work with that version and if it works - for me it is better to investigate the problem with the latest version in another issue

akosyakov commented:
@svor i will, but in 2-3 hours, if you want you can do it before :)

svor commented:
> @svor i will, but in 2-3 hours, if you want you can do it before :)

@akosyakov just tested it and it works fine with 0.16.0. With the latest version 0.18.0 we have some problems (launching a program does not seem to work), I think it isn't related to this PR

elaihau commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -131,7 +135,7 @@ export class PreferenceSchemaProvider extends PreferenceProvider {
 
     protected doSetSchema(schema: PreferenceSchema): PreferenceProviderDataChange[] {
-        const scope = this.getScope();
+        const scope = PreferenceScope.Default;
```
is this something you are going to change in near future? With this change the "scope"s of all prefs will be the same

akosyakov replied:
How can it be anything except default in default preference schema provider?

elaihau replied:
you are right. i misunderstood

elaihau commented on packages/preferences/src/browser/workspace-file-preference-provider.ts:
```diff
@@ -0,0 +1,65 @@
+/********************************************************************************
+ * Copyright (C) 2019 TypFox and others.
```
TypeFox

elaihau commented:
i tested this pr in multi-root workspace context. work well except for the following case:

- add the 1st root folder to an empty multi-root workspace, watch the pref editor.
expected: pref editor should have the 3rd tab added for the folder pref.
actual: no 3rd tab shows up
![Peek 2019-04-26 08-46](https://user-images.githubusercontent.com/37082801/56808611-e5b9ef00-67ff-11e9-826c-57fe1aed9ebd.gif)

It is not necessarily caused by this PR

elaihau commented:
And this case:

Resource still uses workspace preference after being removed from the workspace:

![Peek 2019-04-26 08-59](https://user-images.githubusercontent.com/37082801/56809357-a391ad00-6801-11e9-8c7a-04973531cfb8.gif)

akosyakov commented:
@elaihau thx for the review, found another bug:
<img width="431" alt="Screen Shot 2019-04-26 at 16 28 23" src="https://user-images.githubusercontent.com/3082655/56815202-03f40f00-6841-11e9-9524-7356cfbf9631.png">

duplicate add and remove actions, it is easy to fix

akosyakov commented:
@elaihau What is the content of testdir i cannot reproduce https://github.com/theia-ide/theia/pull/4947#issuecomment-487046371

elaihau commented:
> @elaihau What is the content of testdir i cannot reproduce [#4947 (comment)](https://github.com/theia-ide/theia/pull/4947#issuecomment-487046371)


it has nothing special. the problem can be reproduced with "theia" folder too

![Peek 2019-04-26 10-48](https://user-images.githubusercontent.com/37082801/56816126-e444f280-6810-11e9-9cf2-872a32689214.gif)

elaihau commented:
btw, i also found this one https://github.com/theia-ide/theia/issues/5020, but i don't think it is related to this PR

akosyakov commented:
ok, checking https://github.com/theia-ide/theia/pull/4947#issuecomment-487048059 then, it should work, it would be a bug if there are no events then a preference provider for settings.json is removed

akosyakov commented:
@elaihau i don't think it is caused by this PR, events are reported for removed config files, but since these config files don't contain `editor.fontSize` (it is defined only in workspace settings) then there is no an event for it as well. We could force fire events whenever workspace roots are changed, but maybe it is overkill to cover this corner case?

elaihau commented:
i am ok with issue mentioned in https://github.com/theia-ide/theia/pull/4947#issuecomment-487084726 not fixed, as it is really a corner case.

ariel-bentu commented:
> @ariel-bentu please check whether your launch configuration tests are passing with this PR, they do for me

Hi,
I tested it and it works. the only thing is that VSCode allows comments in the launch.json, and when I used comments and used the api to create a new launch configuration - the configuration saved successfully but the comment was gone.

akosyakov commented:
@ariel-bentu thanks, i will have a look whether it is possible to preserve comments, btw does VS Code do it?

ariel-bentu commented:
> @ariel-bentu thanks, i will have a look whether it is possible to preserve comments, btw does VS Code do it?

I checked and actually, to my surprise, VSCode also looses the comments ... So I guess all is good.
